huaweicloud/services/servicestage/resource_huaweicloud_servicestage_component.go

```diff
@@ -0,0 +1,339 @@
+
+func ResourceComponent() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceComponentCreate,
+		ReadContext:   resourceComponentRead,
+		UpdateContext: resourceComponentUpdate,
+		DeleteContext: resourceComponentDelete,
+
+		Importer: &schema.ResourceImporter{
+			StateContext: resourceComponentImportState,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"region": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+			"application_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.All(
+					validation.StringMatch(regexp.MustCompile(`^[A-Za-z]([\w-]*[A-Za-z0-9])?$`),
+						"The name can only contain letters, digits, underscores (_) and hyphens (-), and the name must"+
+							" start with a letter and end with a letter or digit."),
+					validation.StringLenBetween(2, 64),
+				),
+			},
+			"type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"Webapp", "MicroService", "Common",
+				}, false),
+			},
+			"runtime": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"framework": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"source": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Computed: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"type": {
+							Type:     schema.TypeString,
+							Required: true,
+							ValidateFunc: validation.StringInSlice([]string{
+								"GitHub", "GitLab", "Gitee", "Bitbucket", "package",
+							}, false),
+						},
+						"url": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"authorization": {
+							Type:         schema.TypeString,
+							Optional:     true,
+							Computed:     true,
+							ExactlyOneOf: []string{"source.0.storage_type"},
+						},
+						"storage_type": {
+							Type:     schema.TypeString,
+							Optional: true,
+							Computed: true,
+						},
+					},
+				},
+			},
+			"builder": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Computed: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"organization": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"cluster_id": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"cmd": {
+							Type:     schema.TypeString,
+							Optional: true,
+							Computed: true,
+						},
+						"node_label": {
+							Type:     schema.TypeMap,
+							Optional: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func buildRepoBuilder(params map[string]interface{}) *components.Builder {
```

Comment:
suggest to check the length in this func, so it can be used directly in building createOpts

Reply:
done.